replace histogram with base-2 implementation

brayniac opened the pull request:
Replaces the histogram with a simple base-2 implementation which is opioniated about the values and counter widths. This simplifies usage by converging on a single implementation using atomic counters. Additionally, we get substantial performance benefits by switching to a base-2 implementation.

Similarly the heatmap crate is simplified to converge on a single implementation using atomics.

brayniac commented:
Benchmark results:

```
Original Implementation:

AtomicHistogram/u64/AtomicU64/increment/min/precision/1
                        time:   [7.5811 ns 7.6297 ns 7.6806 ns]
                        thrpt:  [130.20 Melem/s 131.07 Melem/s 131.91 Melem/s]

AtomicHistogram/u64/AtomicU64/increment/max/precision/1
                        time:   [18.881 ns 18.996 ns 19.118 ns]
                        thrpt:  [52.308 Melem/s 52.643 Melem/s 52.962 Melem/s]

New Implementation:

histogram/increment/min time:   [6.1397 ns 6.2107 ns 6.2819 ns]
                        thrpt:  [159.19 Melem/s 161.01 Melem/s 162.87 Melem/s]

histogram/increment/max time:   [6.0612 ns 6.1219 ns 6.1839 ns]
                        thrpt:  [161.71 Melem/s 163.35 Melem/s 164.98 Melem/s]
```

swlynch99 commented on heatmap/src/heatmap.rs:
```diff
@@ -2,36 +2,27 @@
-use rustcommon_time::*;
+use histogram::{Bucket, Histogram};
 
 /// AtomicHeatmaps are concurrent datastructures which store counts for
 /// timestamped values over a configured time range with individual histograms
 /// arranged in a ring buffer. Increments occur in the most recent slice in the
 /// buffer, unless they are newer than that slice may hold. When this happens,
 /// old slices are cleared and reused. This configuration results in a fully
 /// pre-allocated datastructure with concurrent read-write access.
```
This comment should be updated to talk about `Heatmap` now instead of `AtomicHeatmap`.

Also, I suggest having a short intro line and then putting a more detailed description into a separate paragraph since it'll look better in rustdoc.

swlynch99 commented on heatmap/src/heatmap.rs:
```diff
@@ -2,36 +2,27 @@
-    <Count as Atomic>::Primitive: Copy,
-{
+impl Heatmap {
     /// Create a new `AtomicHeatmap` which can store values up and including the
```
Ditto wrt updating `AtomicHeatmap` -> `Heatmap`

brayniac replied:
Fixed

swlynch99 commented on heatmap/src/heatmap.rs:
```diff
@@ -42,27 +33,22 @@ where
-        span: Duration<Nanoseconds<u64>>,
-        resolution: Duration<Nanoseconds<u64>>,
-    ) -> Self {
+    pub fn new(m: u32, r: u32, n: u32, span: Duration, resolution: Duration) -> Self {
```
Could you add some more detailed explanation as to what these parameters mean to the doc comment for this method? Something like
```
- `m` - does this
- `r` - does that
- etc...
```
as is, I have no idea what `m`, `r`, or `n` mean.

brayniac replied:
Good reminder that this needs a builder and a copy paste of the doc from ccommon.

brayniac replied:
Fixed

swlynch99 commented on metrics/src/heatmap.rs:
```diff
@@ -3,16 +3,14 @@
+// use rustcommon_heatmap::AtomicHeatmap;
 
-pub type Heatmap = AtomicHeatmap<u64, AtomicU64>;
+// pub type Heatmap = AtomicHeatmap<u64, AtomicU64>;
```
This commented out code should just be deleted.

brayniac replied:
If this isn't stale, will fix.

swlynch99 commented on metrics/src/lib.rs:
```diff
@@ -74,7 +74,7 @@ mod counter;
 mod lazy;
 
-#[cfg(feature = "heatmap")]
+// #[cfg(feature = "heatmap")]
```
```suggestion
```

brayniac replied:
I think this is stale?

swlynch99 commented on metrics/src/lib.rs:
```diff
@@ -86,7 +86,7 @@ pub use crate::dynmetrics::{DynBoxedMetric, DynPinnedMetric};
 pub use crate::lazy::{Lazy, Relaxed};
 
-#[cfg(feature = "heatmap")]
+// #[cfg(feature = "heatmap")]
```
```suggestion
```

swlynch99 commented on histogram/src/histogram.rs:
```diff
@@ -0,0 +1,430 @@
+
+impl Histogram {
+    #[allow(non_snake_case)]
+    pub fn new(m: u32, r: u32, n: u32) -> Self {
```
This should have a doc comment explaining what `m`, `r`, and `n` mean. People trying to use the heatmap shouldn't have to read the whole implementation just to figure out how to construct it.

brayniac replied:
Done. The builder is more friendly and docs added for those who want to directly construct.

swlynch99 commented on histogram/src/histogram.rs:
```diff
@@ -0,0 +1,430 @@
+    /// Note: if you are reporting on multiple percentiles, it is more efficient
+    /// to use the `percentiles` function to retrieve multiple percentiles in a
+    /// single call.
+    #[allow(clippy::result_unit_err)]
```
This isn't returning `Result<(), Error>`. Does it need to suppress the warning?

brayniac replied:
Fixed.

swlynch99 commented on histogram/src/histogram.rs:
```diff
@@ -0,0 +1,430 @@
+    /// Adds the other `Histogram` to this `Histogram`. Returns an error if
+    /// there are differences in the configurations of both `Histogram`s.
+    #[allow(clippy::result_unit_err)]
+    pub fn add_assign(&self, other: &Self) -> Result<(), ()> {
```
It would be a good idea to create an error type for this. `Result<(), ()>` doesn't play well with libraries like anyhow. Something like this would work (if I remember my thiserror right)

```rs
#[derive(Error)]
#[error("histograms had incompatible parameters")]
struct IncompatibleHistogramError;
```

Also my impression is that most of the literature calls this operation `merge`, up to you on which name you like better.

brayniac replied:
Is there a good equivalent for subtraction to go with merge? I really wanted to use the SubAssign and AddAssign traits, but they're not a good fit for this. 

We already have an error type, so adding something more useful here should be trivial.

swlynch99 replied:
What about `subtract`? It's a verb so IMO it's better than `sub_assign` and the underlying operation is set difference.

brayniac replied:
👍 works for me.

swlynch99 commented on histogram/src/histogram.rs:
```diff
@@ -0,0 +1,430 @@
+    /// Subtracts the other `Histogram` from this `Histogram`. Returns an error
+    /// if there are differences in the configurations of both `Histogram`s.
+    #[allow(clippy::result_unit_err)]
+    pub fn sub_assign(&self, other: &Self) -> Result<(), ()> {
```
Same comment here WRT `Result<(), ()>`

swlynch99 commented on heatmap/src/window.rs:
```diff
@@ -20,7 +20,7 @@ impl<Value, Count> Window<Value, Count> {
     }
 
-    pub fn histogram(&self) -> &Histogram<Value, Count> {
+    pub fn histogram(&self) -> &Histogram {
```
If you have a reference internally it's generally better to pass the lifetime on instead of shortening it 
```suggestion
    pub fn histogram(&self) -> &'a Histogram {
```

brayniac replied:
Fixed.

swlynch99 commented on histogram/Cargo.toml:
```diff
@@ -1,24 +1,16 @@
-description = "Histograms with summary metrics and precision guarantees"
-homepage = "https://github.com/twitter/rustcommon/histogram"
-repository = "https://github.com/twitter/rustcommon"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
```suggestion
```

brayniac replied:
Fixed and added metadata like license =)

swlynch99 reviewed (commented):
Left a bunch of comments, some minor, some a bit more major.

IMO the most important one is that if you're going to use single-letter variable names in parameters then those should be documented thoroughly in the function doc comment. I've skimmed through the entirety of the histogram code and I'm still not sure exactly what `m`, `r`, and `n` mean.

swlynch99 commented on histogram/src/histogram.rs:
```diff
@@ -0,0 +1,430 @@
+    #[allow(non_snake_case)]
+    pub fn new(m: u32, r: u32, n: u32) -> Self {
+        if r <= m || r > n || n > 64 {
+            panic!("invalid histogram config");
```
This should either return an error or it should include a message explaining why the config is invalid.

If you decide to keep the panic then it should also be documented under the `# Panics` header in the doc comment.

swlynch99 commented on histogram/src/histogram.rs:
```diff
@@ -0,0 +1,430 @@
+        }
+
+        let M = 1 << m;
+        let R = (1 << r) - 1;
+        let N = (1 << n) - 1;
```
These will panic in debug mode for `n = 64` and `n = r = 64`. I don't know if this is a huge concern but it's not prevented by the panic above.

brayniac replied:
I *think* that is actually okay as it makes either N or both R and N have the value u64::MAX ?

brayniac replied:
Made this explicit so it no longer panics in debug.

thinkingfish commented:
You can use your base of documentation and edit it as you wish https://github.com/thinkingfish/ccommon/blob/yao/2histogram-doc-md/docs/modules/2-histogram.md

I converted the original rst to markdown using pandoc and fixed most of the styling issues.

brayniac commented:
> You can use your base of documentation and edit it as you wish https://github.com/thinkingfish/ccommon/blob/yao/2histogram-doc-md/docs/modules/2-histogram.md
> 
> I converted the original rst to markdown using pandoc and fixed most of the styling issues.

Thanks! I'll circle back to add some better crate level docs before potentially publishing to crates.io 

For now, I think the rustdocs that I've added are sufficient to get a general sense of how to use this and I would like to get to testing the change in the context of Pelikan
